src/bosh-monitor/lib/bosh/monitor/plugins/json.rb

```diff
@@ -68,39 +73,80 @@ def restart_process(bin)
     def start_process(bin)
       process = Bosh::Monitor::Plugins::DeferrableChildProcess.open(bin)
       process.errback do
-        EventMachine.add_timer(@restart_wait) { restart_process bin }
+        Async do
+          sleep(@restart_wait)
+          restart_process bin
+        end
       end
+      process.run
 
       process
     end
   end
 
-  # EventMachine's DeferrableChildProcess does not give an opportunity
-  # to get the exit status. So we are implementing our own unbind logic to handle the exit status.
-  # This way we can execute our process restart on the err callback (errback).
-  # https://stackoverflow.com/a/12092647
-  class DeferrableChildProcess < EventMachine::Connection
-    include EventMachine::Deferrable
-
-    def initialize
-      super
+  class DeferrableChildProcess
+    def initialize(cmd)
+      @cmd = cmd
       @data = []
+      @errback = []
+      @lock = Mutex.new
+    end
+
+    def run
+      Async do |task|
+        stdin, stdout, stderr, wait_thr = Open3.popen3(@cmd)
+
+        @stdin = Async::IO::Stream.new(Async::IO::Generic.new(stdin))
+        @stdout = Async::IO::Stream.new(Async::IO::Generic.new(stdout))
+        @stderr = Async::IO::Stream.new(Async::IO::Generic.new(stderr))
+        @wait_thr = wait_thr
+
+        task.async do
+          while data = @stdout.read(1)
+            receive_data(data)
+          end
+        end
+
+        task.async do
+          while data = @stderr.read(1)
+            receive_data(data)
+          end
+        end
+
+        # Wait for the process to complete
+        status = @wait_thr.value
+        unless status.success?
+          @errback.each do |errback|
+            errback.call
+          end
+        end
+      rescue => e
+        @errback.each do |errback|
+          errback.call
+        end
+      ensure
+        @stdin.close if @stdin
+        @stdout.close if @stdout
+        @stderr.close if @stderr
+      end
     end
 
     def self.open(cmd)
-      EventMachine.popen(cmd, DeferrableChildProcess)
+      new(cmd)
```

Comment:
Is there a reason to keep `DeferrableChildProcess.open` or could `DeferrableChildProcess.new` be inlined wherever it's called?

Reply:
It could be inlined; part of the reason I didn't is for a couple iterations this method did more, I wanted to minimize interface changes where possible, and it's mocked in tests (mocking constructors is really no different, but always feels worse to me)